docs/querying/api.md

```diff
@@ -336,6 +336,75 @@ $ curl http://localhost:9090/api/v1/label/job/values
 }
 ```
 
+## Querying exemplars
+The following endpoint returns a list of exemplars for a valid PromQL query during a specific time range:
+
+```
+GET /api/v1/query_exemplar
+POST /api/v1/query_exemplar
+```
+
+URL query parameters:
+
+- `query=<string>`: Prometheus expression query string.
+- `start=<rfc3339 | unix_timestamp>`: Start timestamp.
+- `end=<rfc3339 | unix_timestamp>`: End timestamp.
+
+Exemplar storage is not part of Prometheus' TSDB, so this endpoint doesn't allow filtering based on time.
+
+```json
+$ curl -g 'http://localhost:9090/api/v1/query_exemplar?query=test_exemplar_metric_total&start=2020-09-14T15:22:25.479Z&end=020-09-14T15:23:25.479Z'
+{
+    "status": "success",
+    "data": [
+        {
+            "seriesLabels": {
+                "__name__": "test_exemplar_metric_total",
+                "instance": "localhost:8090",
+                "job": "prometheus",
+                "service": "bar"
+            },
+            "exemplars": [
+                {
+                    "labels": {
+                        "traceID": "EpTxMJ40fUus7aGY"
+                    },
+                    "value": 6,
```

Comment:
This like all float values in our APISs this should be a string, +Inf is in theory valid.

Reply:
I'll look at how other endpoints are marshalling data then.

Reply:
Any suggestion for where to make this change? Looks like we just marshal the promql `Result` type:

```
// Result holds the resulting value of an execution or an error
// if any occurred.
type Result struct {
	Err      error
	Value    parser.Value
	Warnings storage.Warnings
}
```

where `Value` is:
```
// Value is a generic interface for values resulting from a query evaluation.
type Value interface {
	Type() ValueType
	String() string
}
```
 and types are:
```
// The valid value types.
const (
	ValueTypeNone   ValueType = "none"
	ValueTypeVector ValueType = "vector"
	ValueTypeScalar ValueType = "scalar"
	ValueTypeMatrix ValueType = "matrix"
	ValueTypeString ValueType = "string"
)
```

The exemplar API endpoint can always return a matrix of exemplars.

Would you suggest something like a new `ExemplarMatrix` type such as: 
```
// Exemplar represents a single data point for a given timestamp.
type Exemplar struct {
        Labels labels.Labels `json:"labels"`
	T int64
	V float64
}

// ExemplarSeries is a stream of exemplars belonging to a metric series.
type ExemplarSeries struct {
	Metric labels.Labels `json:"metric"`
	Exemplars []Exemplars       `json:"exemplars"`
}

// ExemplarMatrix is a slice of ExemplarsSeries that implements sort.Interface and
// has a String method.
type ExemplarMatrix []ExemplarSeries
```

Reply:
https://github.com/prometheus/prometheus/blob/c8f810083d3f6a8b5750186bda13f406980c4218/web/api/v1/api.go#L1621 is the main thing to work off, however that's partially a workaround for our structs not matching the JSON output.
To do this right this time, I'd suggest actually making it a string and not tie our internal representations to the external representations.

Reply:
Not sure if you're suggesting modify/creating a new struct type, or registering something like ```jsoniter.RegisterTypeEncoderFunc("exemplar.Exemplar", marshalExemplarJSON, marshalExemplarJSONIsEmpty)``` that uses something like `Sprintf("%f", exemplar.Value)`.

Reply:
Doc needs to be updated